Introduce function-local settings for executor, expose in c++

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74348
* #74076
* #74186
* __->__ #74012
* #73938
* #73876
* #73875

This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

Commit:
Introduce function-local settings for executor, expose in c++

[ghstack-poisoned]

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor). 

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/runtime/simple_graph_executor_impl.cpp:
```diff
@@ -0,0 +1,43 @@
+    return *execution_plan_;
+  }
+  auto copy = graph->copy();
+  runNooptPassPipeline(copy);
```
lmao why do we call this `Noopt` ? :)

Krovatkin reviewed (approved):
:shipit:

Krovatkin commented on torch/csrc/jit/api/function_impl.h:
```diff
@@ -13,10 +13,14 @@ struct TORCH_API GraphFunction : public Function {
         graph_(std::move(graph)),
-        function_creator_(std::move(function_creator)) {}
+        function_creator_(std::move(function_creator)) {
+    executor_execution_mode_ = executor_execution_mode;
```
why is this initialized inside a c-tor vs doing it like this: `graph_(std::move(graph)),`

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor). 

[ghstack-poisoned]

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor). 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

Commit:
Update on "Introduce function-local settings for executor, expose in c++"


This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).
